### PR TITLE
Issue 6: Handle multiple messages per frame

### LINKED ISSUE #6: Failing to dequeue messages that arrive in rapid succession
https://forum.defold.com/t/extension-websocket-beta-testing/66216/23

### PR DESCRIPTION
**Pull request by JCash**

### CODE REVIEW

File: websocket/src/handshake.cpp [hunk 1]
```diff
@@ -146,7 +146,8 @@ Result ReceiveHeaders(WebsocketConnection* conn)
     conn->m_Buffer[conn->m_BufferSize] = '\0';
 
     // Check if the end of the response has arrived
-    if (conn->m_BufferSize >= 4 && strcmp(conn->m_Buffer + conn->m_BufferSize - 4, "\r\n\r\n") == 0)
+    const char* endtag = strstr(conn->m_Buffer, "\r\n\r\n");
```
Comment (JCash) [inline websocket/src/handshake.cpp:149]: We got the headers _and_ the first messages, so the end tag wasn't at the end.

File: websocket/src/handshake.cpp [hunk 2]
```diff
@@ -218,11 +221,16 @@ Result VerifyHeaders(WebsocketConnection* conn)
...
-            break;
     }
 
+    // The response might contain both the headers, but also (if successful) the first batch of data
+    endtag += 4;
+    uint32_t size = conn->m_BufferSize - (endtag - conn->m_Buffer);
+    conn->m_BufferSize = size;
+    memmove(conn->m_Buffer, endtag, size);
+    conn->m_Buffer[size] = 0;
+    conn->m_HasHandshakeData = conn->m_BufferSize != 0 ? 1 : 0;
```
Comment (JCash) [inline websocket/src/handshake.cpp:232]: Flag that we have lingering data, and store it for later...

File: websocket/src/socket.cpp [hunk 3]
```diff
@@ -42,15 +42,24 @@ dmSocket::Result Send(WebsocketConnection* conn, const char* buffer, int length,
     }
     if (out_sent_bytes)
         *out_sent_bytes = total_sent_bytes;
+
+    DebugPrint(2, "Sent buffer:", buffer, length);
```
Comment (JCash) [inline websocket/src/socket.cpp:46]: For future debugging, we now have `websocket.debug = ...` that will enable more debug info.

File: websocket/src/websocket.cpp [hunk 4]
```diff
@@ -380,7 +436,7 @@ static dmExtension::Result WebsocketAppInitialize(dmExtension::AppParams* params
     return dmExtension::RESULT_OK;
 }
 
-static dmExtension::Result WebsocketInitialize(dmExtension::Params* params)
+static dmExtension::Result Initialize(dmExtension::Params* params)
```
Comment (JCash) [inline websocket/src/websocket.cpp:444]: They're already in the `dmWebSocket` namespace, so it feels redundant to name it "dmWebSocket::WebsocketInitialize`

File: websocket/src/websocket.cpp [hunk 5]
```diff
@@ -391,19 +447,27 @@ static dmExtension::Result WebsocketInitialize(dmExtension::Params* params)
...
 }
 
-static dmExtension::Result WebsocketFinalize(dmExtension::Params* params)
+static dmExtension::Result Finalize(dmExtension::Params* params)
 {
     return dmExtension::RESULT_OK;
 }
 
-static dmExtension::Result WebsocketOnUpdate(dmExtension::Params* params)
+Result PushMessage(WebsocketConnection* conn, int length)
```
Comment (JCash) [inline websocket/src/websocket.cpp:467]: Data is already appended to the connection buffer.

File: websocket/src/wslay_callbacks.cpp [hunk 6]
```diff
@@ -82,6 +82,15 @@ ssize_t WSL_RecvCallback(wslay_event_context_ptr ctx, uint8_t *buf, size_t len,
...
     int r = -1; // received bytes if >=0, error if < 0
 
+    if (conn->m_HasHandshakeData)
+    {
+        r = conn->m_BufferSize;
+        memcpy(buf, conn->m_Buffer, r);
+        conn->m_BufferSize = 0;
+        conn->m_HasHandshakeData = 0;
+        return r;
+    }
```
Comment (JCash) [inline websocket/src/wslay_callbacks.cpp:92]: The "trick". When `wslay`wants to read from the socket, we first return any lingering data from the handshake phase.

File: websocket/src/wslay_callbacks.cpp [hunk 7]
```diff
@@ -117,16 +126,23 @@ ssize_t WSL_SendCallback(wslay_event_context_ptr ctx, const uint8_t *data, size_
...
+        if ((conn->m_BufferSize + arg->msg_length) >= conn->m_BufferCapacity)
+        {
+            conn->m_BufferCapacity = conn->m_BufferSize + arg->msg_length + 1;
+            conn->m_Buffer = (char*)realloc(conn->m_Buffer, conn->m_BufferCapacity);
+        }
+        // append to the end of the buffer
+        memcpy(conn->m_Buffer + conn->m_BufferSize, arg->msg, arg->msg_length);
+        conn->m_BufferSize += arg->msg_length;
+
+        PushMessage(conn, arg->msg_length);
```
Comment (JCash) [inline websocket/src/wslay_callbacks.cpp:144]: Previously, we only stored 1 message per frame, but here we get several in a single frame.

**Review by britzl — APPROVED** [review]

**Pull request merged**